arroyo-sql/src/lib.rs

```diff
@@ -439,55 +359,249 @@ pub async fn parse_and_get_program(
     query: &str,
     mut schema_provider: ArroyoSchemaProvider,
     config: SqlConfig,
-) -> Result<(Program, Vec<SqlSource>)> {
+) -> Result<(Program, Vec<i64>)> {
     let query = query.to_string();
 
     if query.trim().is_empty() {
         bail!("Query is empty");
     }
 
     tokio::spawn(async move {
-        // parse the SQL
-        let plan = get_plan_from_query(&query, &mut schema_provider)?;
-        get_program_from_plan(config, schema_provider, &plan)
+        let mut sql_program_builder = SqlProgramBuilder {
+            schema_provider: &mut schema_provider,
+        };
+        let outputs = sql_program_builder.plan_query(&query)?;
+        let mut sql_pipeline_builder =
+            SqlPipelineBuilder::new(sql_program_builder.schema_provider);
+        for output in outputs {
+            sql_pipeline_builder.insert_table(output)?;
+        }
+        let mut plan_graph = PlanGraph::new(config.clone());
+        let last_output = sql_pipeline_builder
+            .output_nodes
+            .last()
+            .ok_or_else(|| anyhow!("No output nodes"))?;
+
+        // If the last output is not a sink, add the sink passed in.
+        if !matches!(last_output, SqlOperator::Sink(..)) {
+            let non_sink_output = sql_pipeline_builder.output_nodes.pop().unwrap();
+            let struct_def = non_sink_output.return_type();
+            let sink = SqlOperator::Sink(
+                "default_sink".to_string(),
+                SqlSink {
+                    id: None,
+                    struct_def,
+                    sink_config: config.sink.clone(),
+                },
+                Box::new(non_sink_output),
+            );
+            sql_pipeline_builder.output_nodes.push(sink);
+        }
+
+        for output in sql_pipeline_builder.output_nodes.into_iter() {
+            plan_graph.add_sql_operator(output);
+        }
+        get_program(plan_graph, sql_program_builder.schema_provider.clone())
     })
     .await
     .map_err(|_| anyhow!("Something went wrong"))?
 }
 
-fn get_plan_from_query(
-    query: &str,
-    schema_provider: &mut ArroyoSchemaProvider,
-) -> Result<LogicalPlan> {
-    // parse the SQL
-    let dialect = PostgreSqlDialect {};
-    let mut ast = Parser::parse_sql(&dialect, &query)
-        .map_err(|e| match e {
-            ParserError::TokenizerError(s) | ParserError::ParserError(s) => anyhow!(s),
-            ParserError::RecursionLimitExceeded => anyhow!("recursion limit"),
-        })
-        .with_context(|| "parse_failure")?;
-    if ast.len() == 0 {
-        bail!("Query is empty");
+struct SqlProgramBuilder<'a> {
+    schema_provider: &'a mut ArroyoSchemaProvider,
+}
+
+impl<'a> SqlProgramBuilder<'a> {
+    fn plan_query(&mut self, query: &str) -> Result<Vec<Table>> {
+        let dialect = PostgreSqlDialect {};
+        let mut outputs = Vec::new();
+        for statement in Parser::parse_sql(&dialect, query).unwrap() {
```

Comment:
This unwrap seems to remove error handling for parse errors -- is that what we want?

For example, if I accidentally use the wrong quote terminator, previously I would have gotten this error in the frontend:

```
Unterminated string literal at Line: 11, Column 22
```

but now I just get a generic error.